presto-hive/src/main/java/io/prestosql/plugin/hive/security/SqlStandardAccessControlMetadata.java

```diff
@@ -84,6 +87,62 @@ private static void checkRoleIsNotReserved(String role)
         return ImmutableSet.copyOf(metastore.listRoles());
     }
 
+    @Override
+    public Set<RoleGrant> listAllRoleGrants(ConnectorSession session, Optional<Set<String>> roles, Optional<Set<String>> grantees, OptionalLong limit)
+    {
+        /*
+         * The Hive Metadata interface provides us with two ways to retrieve role grants:
+         * 1. principals by single role
+         * 2. roles by single principal
+         *
+         * We check the passed predicates to decide which path is most optimal.
+         */
+        if (roles.isPresent() && grantees.isPresent()) {
+            if (roles.get().size() <= grantees.get().size() * 2) {
+                // 2x because for the grantees we do not know the grantee type (ROLE or USER)
+                grantees = Optional.empty();
+            }
+            else {
+                roles = Optional.empty();
+            }
+        }
+
+        ImmutableSet.Builder<RoleGrant> builder = ImmutableSet.builder();
+        long count = 0;
+        if (grantees.isPresent()) {
+            TOP:
+            for (String grantee : grantees.get()) {
+                for (PrincipalType type : new PrincipalType[]{USER, ROLE}) {
+                    for (RoleGrant grant : metastore.listRoleGrants(new HivePrincipal(type, grantee))) {
+                        if (limit.isPresent() && count >= limit.getAsLong()) {
+                            break TOP;
+                        }
+                        // Filter out the "public" role since it is not explicitly granted in Hive.
+                        if ("public".equals(grant.getRoleName())) {
```

Comment:
use some constant for `"public"`. There should be some already.

Reply:
The trick is that we absolutely need to honor the limit the first time it hits. Each loop here is a roundtrip to the metastore (that's why the logic is a bit convoluted with the break label).

I could pass the builder to a method and that method returns a boolean to indicate whether we had hit the limit or not (so that the outer loops knows it's done).

Not sure that is much prettier. At least with the behavior is explicit. Lemme mull over that a bit more.

Reply:
On "public". Honestly, none that I could find. The "public" string is all over the place in the Hive connector. There're private constants iniThriftMetastoreUtil, I could make those public.

Reply:
Added a constant to this class at least.

Reply:
Cleaned up the logic too, extracted methods, etc.